File: MORE/LPDAAC/MOD/MOD14.A2023221.0750.061.2023221151516.hdf.py
"""
This example code illustrates how to access and visualize a LP DAAC MODIS swath
file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MOD021KM.A2023223.0455.061.2023223132259.hdf.py

The HDF files must be in your current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2023-08-17
"""
import os
import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap
import numpy as np

USE_NETCDF4 = False
FILE_NAME = 'MOD14.A2023221.0750.061.2023221151516.hdf'
GEO_FILE_NAME = 'MOD03.A2023221.0750.061.2023221131337.hdf'
DATAFIELD_NAME = 'fire mask'

from pyhdf.SD import SD, SDC
hdf = SD(FILE_NAME, SDC.READ)

# Read dataset.
data2D = hdf.select(DATAFIELD_NAME)
data = data2D[:,:].astype(np.double)

hdf_geo = SD(GEO_FILE_NAME, SDC.READ)

# Read geolocation dataset from MOD03 product.
lat = hdf_geo.select('Latitude')
latitude = lat[:,:]
lon = hdf_geo.select('Longitude')
longitude = lon[:,:]
        
# Retrieve attributes.
attrs = data2D.attributes(full=1)
lna=attrs["legend"]
long_name = lna[0]
vra=attrs["valid_range"]
valid_min = vra[0][0]        
valid_max = vra[0][1]        

        
invalid = np.logical_or(data > valid_max,
                        data < valid_min)
data[invalid] = np.nan
data = np.ma.masked_array(data, np.isnan(data))

# Find middle location.
lat_m = np.nanmean(latitude)
lon_m = np.nanmean(longitude)
lat_min = np.nanmin(latitude)
lat_max = np.nanmax(latitude)
lon_min = np.nanmin(longitude)
lon_max = np.nanmax(longitude)

# Render the plot in a lambert equal area projection.
m = Basemap(projection='laea', resolution='l', lat_ts=65,
            lat_0=lat_m, lon_0=lon_m,
            width=3000000, height=2500000)
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(lat_min, lat_max, 10), labels=[1, 0, 0, 0])
m.drawmeridians(np.arange(lon_min, lon_max, 10), labels=[0, 0, 0, 1])
m.scatter(longitude, latitude, c=data, s=1, cmap=plt.cm.jet,
           edgecolors=None, linewidth=0, latlon=True)
cb = m.colorbar()

# Exagerate dot size for fire locations.
idx = (data >= 7)
m.scatter(longitude[idx], latitude[idx], c=data[idx], s=3, cmap=plt.cm.jet,
           edgecolors=None, linewidth=0, latlon=True)
units = long_name[-77:]
cb.set_label(units, fontsize=8)

basename = os.path.basename(FILE_NAME)
plt.title('{0}\n{1}'.format(basename, DATAFIELD_NAME))
fig = plt.gcf()
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)

File: MORE/LPDAAC/MOD/MOD14_A2023221_0750_061_2023221151516_hdf.m
% This example code illustrates how to access and visualize LP DAAC MODIS Swath
% file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, please 
% use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA HDF/HDF-EOS data 
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org or 
% post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage: save this script and run (without .m at the end)
% 
% $matlab -nosplash -nodesktop -r MOD14_A2023221_0750_061_2023221151516_hdf
%
% Tested under: MATLAB R2023a
% Last updated: 2023-08-17

import matlab.io.hdfeos.*
import matlab.io.hdf4.*

% Read lat and lon data.
GEO_FILE_NAME='MOD03.A2023221.0750.061.2023221131337.hdf';
SWATH_NAME='MODIS_Swath_Type_GEO';

% Open the HDF-EOS2 swath File
file_id = sw.open(GEO_FILE_NAME, 'rdonly');

% Open swath.
swath_id = sw.attach(file_id, SWATH_NAME);

% Read lat and lon data.
lon = sw.readField(swath_id, 'Longitude', [], [], []);
lat = sw.readField(swath_id, 'Latitude', [], [], []);

% Detach from swath.
sw.detach(swath_id);
sw.close(file_id);

lon=double(lon);
lat=double(lat);

% Set file name and dataset name.
FILE_NAME = 'MOD14.A2023221.0750.061.2023221151516.hdf';
DATAFIELD_NAME = 'fire mask';

% Read data.
SD_id = sd.start(FILE_NAME, 'rdonly');
sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
sds_id = sd.select(SD_id, sds_index);

% Read attributes from data field.
data1 = sd.readData(sds_id);

% Get legend attribute.
long_name_index = sd.findAttr(sds_id, 'legend');
long_name = sd.readAttr(sds_id, long_name_index);

% Read valid_range attribute.
range_index = sd.findAttr(sds_id, 'valid_range');
range = sd.readAttr(sds_id, range_index);

% Terminate access.
sd.endAccess(sds_id);

% Close file.
sd.close(SD_id);

data = double(data1);

% Replace fill value with NaN.
data(data > range(2)) = NaN;


% Create figure.
f = figure('Name', FILE_NAME, ...
           'Renderer', 'zbuffer', ...
           'Position', [0,0,800,600], ...
           'visible', 'off');

% Set the map parameters.
lon_c = mean(mean(lon));
lat_c = mean(mean(lat));
latlim = ceil(max(max(lat))) - floor(min(min(lat)));

% FlatLimit will give us a zoom-in effect in Ortho projection.
axesm ('ortho', 'Frame', 'on', 'Grid', 'on', ...
       'FLatLimit', [-Inf, latlim], ...
       'origin', [lat_c, lon_c])
mlabel('equator')
plabel(0); 
plabel('fontweight','bold')
surfm(lat, lon, data);

% Highlight fire data.
idx = (data > 6);
scatterm(lat(idx), lon(idx), 3, data(idx));

% Load the coastlines data file.                                
coast = load('coastlines.mat');

% Plot coastlines in black color.                                
plotm(coast.coastlat,coast.coastlon, 'k')

colormap('Jet');
h = colorbar();
set(get(h, 'title'), 'string', long_name(length(long_name)-76:end));

% Set the title using long name attribute.
title({FILE_NAME; ...
      ['FIELD: ', DATAFIELD_NAME ]}, ...
      'FontSize',10,'FontWeight','bold');
saveas(f, [FILE_NAME '.m.png']);
exit;


